server/routes/user.py

```diff
@@ -1,22 +1,98 @@
+import datetime, uuid, jwt
 
+# Just for development purpose i.e. to display all users in the database
+@app.route("/users/display", methods=["GET"])
+def display():
+    users = User.query.all()
+    output = []
 
+    for user in users:
+        user_data = {}
+        user_data["id"] = user.id
+        user_data["username"] = user.username
+        user_data["email"] = user.email
+        output.append(user_data)
+
+    return jsonify({"users": output})
+
+
+# Login Endpoint
 @app.route("/users/login", methods=["GET", "POST"])
 def login():
-    return dedent(
-        """
-        <h1>USER LOGIN API</h1>
-        <a href="/">Back to home</a>
-    """
+    body = request.json
+    # return jsonify({"user": auth})
+
+    if not body:
+        return make_response(
+            "Could not verify",
+            401,
+            {"WWW-Authenticate": 'Basic realm="Login required!"'},
+        )
+
+    user = User.query.filter_by(email=body["email"]).first()
+
+    if not user:
+        return make_response(
+            "Could not verify",
+            401,
+            {"WWW-Authenticate": 'Basic realm="Login required!"'},
+        )
+
+    if bcrypt.check_password_hash(user.password, body["password"]):
+        token = jwt.encode(
+            {
+                "id": user.id,
+                "exp": datetime.datetime.utcnow() + datetime.timedelta(minutes=30),
+            },
+            app.config["SECRET_KEY"],
+        )
+        return jsonify({"token": token})
+
+    return make_response(
+        "Could not verify", 401, {"WWW-Authenticate": 'Basic realm="Login required!"'}
     )
 
 
-@app.route("/users/signup", methods=["GET", "POST"])
+@app.route("/users/register", methods=["POST"])
 def signup():
-    return dedent(
-        """
-        <h1>USER SIGNUP API</h1>
-        <a href="/">Back to home</a>
     """
-    )
+    We would be expecting a request body of this type:
+    {
+        'username' : "abc",
+        'password': "abc",
+        'email': "abc@abc"
+    }
+    """
+    try:
+        body = request.json
+        if body:
+            username = body["username"]
+            password = body["password"]
+            email = body["email"]
+            hashed_password = bcrypt.generate_password_hash(password)
+            user = User(
+                id=str(uuid.uuid4()),
+                username=username,
+                password=hashed_password,
+                email=email,
+            )
+            db.session.add(user)
+            db.session.commit()
+
+            current_app.logger.info(
+                f"<AUTH> Adding new user: {user.username}, email: {user.email}"
+            )
+            return make_response(
+                "<h2>Welcome to Journal Policy Tracker.</h2><br>Have a great time.".format(
+                    user.username
+                ),
+                201,
+            )
+    except:
+        current_app.logger.error("<REGISTRATION> Unable to parse POST request.")
```

Comment:
I would avoid "Pokémon-exceptions" (i.e. exceptions that catch everything, so "gotta catch'em all") as much as possible, unless it is to log and forward it up again, This "forward" should be done by simply calling `raise` (not `raise <exception>`, as this would overwrite the call stack):

```suggestion
    except:
        current_app.logger.error("<REGISTRATION> Unable to parse POST request.")
        raise
```

In this specific case, a failure on register becomes difficult to grasp what happened (was it a illformed JSON? Was it a typo in a method call?).

Reply:
Is it okay now @JPTIZ?

Reply:
Way better, now I know what error I am getting :rofl: 

Apparently the user table isn't being generated automatically with the application start, do you have any clues? Maybe I'm missing some operation.

Logs:

```
[2021-07-28 20:15:25,296] ERROR in user: <REGISTRATION> Unable to parse POST request.                                                                                                                                                          
127.0.0.1 - - [28/Jul/2021 20:15:25] "POST /users/register HTTP/1.1" 500 -                                                                                                                                                                     Traceback (most recent call last):                                                                                                                                                                                                             
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1771, in _execute_context                                                                                
    self.dialect.do_execute(                                                                                                                                                                                                                     File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 717, in do_execute                                                                                    
    cursor.execute(statement, parameters)                                                                                                                                                                                                      
sqlite3.OperationalError: no such table: user                                                                                                                                                                                                  
                                                                                                                                                                                                                                               
The above exception was the direct cause of the following exception:                                                                                                                                                                           
                                                                                                                                                                                                                                               
Traceback (most recent call last):                                                                                                                                                                                                             
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 2464, in __call__                                                                                                     
    return self.wsgi_app(environ, start_response)                                                                                                                                                                                              
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 2450, in wsgi_app                                                                                                     
    response = self.handle_exception(e)                                                                                                                                                                                                        
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 1867, in handle_exception                                                                                             
    reraise(exc_type, exc_value, tb)                                                                                                                                                                                                           
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise                                                                                                    
    raise value                                                                                                                                                                                                                                
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 2447, in wsgi_app                                                                                                     
    response = self.full_dispatch_request()                                                                                                                                                                                                    
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 1952, in full_dispatch_request                                                                                        
    rv = self.handle_user_exception(e)                                                                                                                                                                                                         
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 1821, in handle_user_exception                                                                                        
    reraise(exc_type, exc_value, tb)
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/jptiz/.cache/pypoetry/virtualenvs/codecompl-SX4IdcVr-py3.9/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/jptiz/dev/codeisscience/codecompliance-backend/server/routes/user.py", line 84, in signup
    db.session.commit()
  File "<string>", line 2, in commit
```

Reply:
Ok, creating the database with...

```
$ poetry shell
$ python
>>> from server import db
>>> db.create_all()
```

Solved my issue :)